[ONNX] Improve diagnostics performance

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99944
* #99668
* __->__ #99936

Summary
- Do not call `fx_graph_module.print_readable` when recording `fx.GraphModule` function argument diagnostics. 
- Cache `inspect.getsourcelines` results.

BowenBao committed:
[ONNX] Improve diagnostics performance

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Improve diagnostics performance"




[ghstack-poisoned]

BowenBao commented:
I'm inclined to include a sanity test like below, but didn't due to concerns over its flakiness. Ideas are welcomed.
```python
    def test_export_remains_efficient_with_diagnostics(self):
        model_name = "gpt2"
        # Download pytorch model
        model = transformers.AutoModel.from_pretrained(model_name)
        tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)
        inputs = tokenizer("Hello world!", return_tensors="pt")

        start_time = time.time()
        with common_utils.TemporaryFileName() as path:
            torch.onnx.dynamo_export(model, **inputs).save(path)
        elapsed_time = time.time() - start_time
        time_threshold_in_seconds = 15.0
        self.assertTrue(
            elapsed_time < time_threshold_in_seconds,
            (
                f"Exporting GPT2 model took too long! "
                f"{elapsed_time} seconds > {time_threshold_in_seconds} seconds."
                f"This is a sanity check that `torch.onnx.dynamo_export` remains "
                f"reasonably efficient with all the diagnostics and analysis enabled. "
                f"The time constraint is loosely set such that the test should pass "
                f"on most machines."
            ),
        )
```

justinchuby commented on torch/onnx/_internal/diagnostics/infra/utils.py:
```diff
@@ -41,13 +43,24 @@ def python_call_stack(frames_to_skip: int = 0, frames_to_log: int = 16) -> _infr
     return stack
 
 
+@functools.lru_cache()
```
Do we need to set a size for the cache?

BowenBao replied:
I think it's by default 128? Seems good enough as it made this function disappear in profiler.

justinchuby replied:
Good catch I didn’t realize 128 was the default

justinchuby commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -46,12 +46,12 @@ def format_argument(obj: Any) -> str:
 @_format_argument.register
 def _torch_nn_module(obj: torch.nn.Module) -> str:
-    return f"{obj.__class__.__name__}"
+    return f"torch.nn.Module({obj.__class__.__name__})"
```
I would add a comment on these change to remind future readers lest they change them back

BowenBao replied:
added a comment section for 'read before modify'

justinchuby commented:
Curious on the speed gain?

abock commented:
I was just about to suggest we include GPT-2 as some kind of nodes/s baseline test with a large tolerance.

```
   Uncached stack:     36 nodes/s  (1x)    ← Baseline
        LRU stack:  1,208 nodes/s  (33.5x) ← This PR  🎉
    Omitted stack:  2,827 nodes/s  (78.5x) ← Not helpful
```

This PR is a good balance and we should merge it as-is (unless you wan to add the test), but we may want to further only gather stack info under trace/debug level for another ~2x bump later?

abock commented:
@justinchuby for context, I stuck `tqdm` around the node loop in `_export_fx_node_to_onnxscript` for my own amusement.

abock commented:
```python
import torch
import transformers

torch.onnx.dynamo_export(
    transformers.GPT2Model.from_pretrained("gpt2"),
    **transformers.GPT2Tokenizer.from_pretrained("gpt2")(
        "Tokenize me",
        return_tensors="pt",
    ),
).save("gpt2.onnx")
```

BowenBao commented:
@abock thanks for posting speed gain. Yep I think it should be configurable through api. We'd want export to be fast so any perf heavy diagnosing should hide behind it. Merging after adding comments per @justinchuby 's suggestion.

BowenBao committed:
Update on "[ONNX] Improve diagnostics performance"


Summary
- Do not call `fx_graph_module.print_readable` when recording `fx.GraphModule` function argument diagnostics. 
- Cache `inspect.getsourcelines` results.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge

BowenBao committed:
Update on "[ONNX] Improve diagnostics performance"


Summary
- Do not call `fx_graph_module.print_readable` when recording `fx.GraphModule` function argument diagnostics. 
- Cache `inspect.getsourcelines` results.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
